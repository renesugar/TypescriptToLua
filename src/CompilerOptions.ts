import * as ts from "typescript";

export interface CompilerOptions extends ts.CompilerOptions {
    noHeader?: boolean;
    luaTarget?: LuaTarget;
    luaLibImport?: LuaLibImportKind;
    noHoisting?: boolean;
}

export enum LuaLibImportKind {
    None = "none",
    Always = "always",
    Inline = "inline",
    Require = "require",
}

export enum LuaTarget {
    Lua51 = "5.1",
    Lua52 = "5.2",
    Lua53 = "5.3",
    LuaJIT = "jit",
}
