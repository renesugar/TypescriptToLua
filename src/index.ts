export {parseCommandLine} from "./CommandLineParser";
export {compile, compileFilesWithOptions, transpileString, watchWithOptions} from "./Compiler";
export {CompilerOptions, LuaLibImportKind, LuaTarget,} from "./CompilerOptions";
export {LuaLibFeature,} from "./LuaLib";
export {LuaTranspiler,} from "./LuaTranspiler";
