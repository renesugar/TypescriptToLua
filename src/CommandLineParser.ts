import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";

import {CompilerOptions, LuaTarget, LuaLibImportKind} from "./CompilerOptions";

export type CLIParseResult = ParseResult<ParsedCommandLine>;

type ParseResult<T> =
    { isValid: true; result: T }
    | { isValid: false, errorMessage: string};

type ArgumentParseResult<T> =
    { isValid: true; result: T; increment?: number }
    | { isValid: false, errorMessage: string };

interface ParsedCommandLine extends ts.ParsedCommandLine {
    options: CompilerOptions;
}

interface BaseCLIOption {
    aliases: string[];
    describe: string;
    type: string;
}

interface CLIOption<T> extends BaseCLIOption {
    choices: T[];
    default: T;
}

const optionDeclarations: {[key: string]: CLIOption<any>} = {
    luaLibImport: {
        choices: [LuaLibImportKind.Inline, LuaLibImportKind.Require, LuaLibImportKind.Always, LuaLibImportKind.None],
        default: LuaLibImportKind.Inline,
        describe: "Specifies how js standard features missing in lua are imported.",
        type: "enum",
    } as CLIOption<string>,
    luaTarget: {
        aliases: ["lt"],
        choices: [LuaTarget.LuaJIT, LuaTarget.Lua53, LuaTarget.Lua52, LuaTarget.Lua51],
        default: LuaTarget.LuaJIT,
        describe: "Specify Lua target version.",
        type: "enum",
    } as CLIOption<string>,
    noHeader: {
        default: false,
        describe: "Specify if a header will be added to compiled files.",
        type: "boolean",
    } as CLIOption<boolean>,
    noHoisting: {
        default: false,
        describe: "Disables hoisting.",
        type: "boolean",
    } as CLIOption<boolean>,
};

export const { version } = require("../package.json");

const helpString =
    `Version ${version}\n` +
    "Syntax:   tstl [options] [files...]\n\n" +

    "Examples: tstl path/to/file.ts [...]\n" +
    "          tstl -p path/to/tsconfig.json\n\n" +

    "In addition to the options listed below you can also pass options\n" +
    "for the typescript compiler (For a list of options use tsc -h).\n" +
    "Some tsc options might have no effect.";

/**
 * Parse the supplied arguments.
 * The result will include arguments supplied via CLI and arguments from tsconfig.
 */
export function parseCommandLine(args: string[]): CLIParseResult
{
    let commandLine = ts.parseCommandLine(args);

    // Run diagnostics to check for invalid tsc options
    const diagnosticsResult = runTsDiagnostics(commandLine);
    if (diagnosticsResult.isValid === false) {
        return diagnosticsResult;
    }

    // This will add TS and TSTL options from a tsconfig
    const configResult = readTsConfig(commandLine);
    if (configResult.isValid === true) {
        commandLine = configResult.result;
    } else {
        return { isValid: false, errorMessage: configResult.errorMessage };
    }

    // Run diagnostics to check for invalid tsconfig
    const diagnosticsResult2 = runTsDiagnostics(commandLine);
    if (diagnosticsResult2.isValid === false) {
        return diagnosticsResult2;
    }

    // Merge TSTL CLI options in (highest priority) will also set defaults if none specified
    const tstlCLIResult = parseTSTLOptions(commandLine, args);
    if (tstlCLIResult.isValid === true) {
        commandLine = tstlCLIResult.result;
    } else {
        return { isValid: false, errorMessage: tstlCLIResult.errorMessage };
    }

    if (commandLine.options.project && !commandLine.options.rootDir) {
        commandLine.options.rootDir = path.dirname(commandLine.options.project);
    }

    if (!commandLine.options.rootDir) {
        commandLine.options.rootDir = process.cwd();
    }

    if (!commandLine.options.outDir) {
        commandLine.options.outDir = commandLine.options.rootDir;
    }

    return { isValid: true, result: commandLine as ParsedCommandLine };
}

export function getHelpString(): string {
    let result = helpString + "\n\n";

    result += "Options:\n";
    for (const optionName in optionDeclarations) {
        const option = optionDeclarations[optionName];
        const aliasStrings = option.aliases
            ? option.aliases.map(a => "-" + a)
            : [];

        const optionString = aliasStrings.concat(["--" + optionName]).join("|");

        const parameterDescribe = option.choices
            ? option.choices.join("|")
            : option.type;

        const spacing = " ".repeat(Math.max(1, 45 - optionString.length - parameterDescribe.length));

        result += `\n ${optionString} <${parameterDescribe}>${spacing}${option.describe}\n`;
    }

    return result;
}

function readTsConfig(parsedCommandLine: ts.ParsedCommandLine): CLIParseResult
{
    const options = parsedCommandLine.options;

    // Load config
    if (options.project) {
        const findProjectPathResult = findConfigFile(options);
        if (findProjectPathResult.isValid === true) {
            options.project = findProjectPathResult.result;
        } else {
            return { isValid: false, errorMessage: findProjectPathResult.errorMessage };
        }

        const configPath = options.project;
        const configContents = fs.readFileSync(configPath).toString();
        const configJson = ts.parseConfigFileTextToJson(configPath, configContents);
        const parsedJsonConfig = ts.parseJsonConfigFileContent(
            configJson.config,
            ts.sys,
            path.dirname(configPath),
            options
        );

        for (const key in parsedJsonConfig.raw) {
            const option = optionDeclarations[key];
            if (option !== undefined) {
                const value = readValue(parsedJsonConfig.raw[key], option.type, key);
                if (option.choices) {
                    if (option.choices.indexOf(value) < 0) {
                        return {
                            isValid: false,
                            errorMessage: `Unknown ${key} value '${value}'.\nAccepted values: ${option.choices}`,
                        };
                    }
                }

                parsedJsonConfig.options[key] = value;
            }
        }
        return { isValid: true, result: parsedJsonConfig };
    }
    return { isValid: true, result: parsedCommandLine };
}

function parseTSTLOptions(commandLine: ts.ParsedCommandLine, args: string[]): CLIParseResult {
    const result = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith("--")) {
            const argumentName = args[i].substr(2);
            const option = optionDeclarations[argumentName];
            if (option) {
                const argumentResult = getArgumentValue(argumentName, i, args);
                if (argumentResult.isValid === true) {
                    result[argumentName] = argumentResult.result;
                    // Skip value from being considered as option
                    i += argumentResult.increment !== undefined ? argumentResult.increment : 1;
                } else {
                    return { isValid: false, errorMessage: argumentResult.errorMessage };
                }
            }
        } else if (args[i].startsWith("-")) {
            const argument = args[i].substr(1);
            let argumentName: string;
            for (const key in optionDeclarations) {
                if (optionDeclarations[key].aliases && optionDeclarations[key].aliases.indexOf(argument) >= 0) {
                    argumentName = key;
                    break;
                }
            }

            if (argumentName) {
                const argumentResult = getArgumentValue(argumentName, i, args);
                if (argumentResult.isValid === true) {
                    result[argumentName] = argumentResult.result;
                    // Skip value from being considered as option
                    i += argumentResult.increment !== undefined ? argumentResult.increment : 1;
                } else {
                    return { isValid: false, errorMessage: argumentResult.errorMessage };
                }
            }
        }
    }
    for (const option in result) {
        commandLine.options[option] = result[option];
    }
    // Add defaults if not set
    const defaultOptions = getDefaultOptions();
    for (const option in defaultOptions) {
        if (!commandLine.options[option]) {
            commandLine.options[option] = defaultOptions[option];
        }
    }
    return { isValid: true, result: commandLine };
}

function getArgumentValue(
    argumentName: string,
    argumentIndex: number,
    args: string[]
): ArgumentParseResult<string | boolean>
{
    const option = optionDeclarations[argumentName];
    const argument = args[argumentIndex + 1];

    if (option.type === "boolean" && (argument === undefined || argument.startsWith("-"))) {
        // Set boolean arguments without supplied value to true
        return { isValid: true, result: true, increment: 0 };
    }

    if (argument === undefined) {
        return { isValid: false, errorMessage: `Missing value for parameter ${argumentName}`};
    }

    const value = readValue(argument, option.type, argumentName);

    if (option.choices) {
        if (option.choices.indexOf(value) < 0) {
            return {
                isValid: false,
                errorMessage: `Unknown ${argumentName} value '${value}'. Accepted values are: ${option.choices}`,
            };
        }
    }

    return { isValid: true, result: value };
}

function readValue(valueString: string, valueType: string, parameterName: string): string | boolean {
    if (valueType === "boolean") {
        return valueString === "true" || valueString === "t"
            ? true
            : false;
    } else if (valueType === "enum") {
        return valueString.toLowerCase();
    } else {
        return valueString;
    }
}

function getDefaultOptions(): CompilerOptions {
    const options: CompilerOptions = {};

    for (const optionName in optionDeclarations) {
        if (optionDeclarations[optionName].default !== undefined) {
            options[optionName] = optionDeclarations[optionName].default;
        }
    }

    return options;
}

/** Check the current state of the ParsedCommandLine for errors */
function runTsDiagnostics(commandLine: ts.ParsedCommandLine): ParseResult<boolean> {
    // Remove files that dont exist
    commandLine.fileNames = commandLine.fileNames.filter(file => fs.existsSync(file) || fs.existsSync(file + ".ts"));

    const tsInvalidCompilerOptionErrorCode = 5023;
    if (commandLine.errors.length !== 0) {
        // Generate a list of valid option names and aliases
        const optionNames: string[] = [];
        for (const key of Object.keys(optionDeclarations)) {
            optionNames.push(key);
            const alias = optionDeclarations[key].aliases;
            if (alias) {
                if (typeof alias === "string") {
                    optionNames.push(alias);
                } else {
                    optionNames.push(...alias);
                }
            }
        }

        for (const err of commandLine.errors) {
            // Ignore errors caused by tstl specific compiler options
            if (err.code === tsInvalidCompilerOptionErrorCode) {
                let ignore = false;
                for (const optionName of optionNames) {
                    if (err.messageText.toString().indexOf(optionName) !== -1) {
                        ignore = true;
                        break;
                    }
                }

                if (!ignore) {
                    return { isValid: false, errorMessage: `error TS${err.code}: ${err.messageText}`};
                }
            }
        }
    }

    return { isValid: true, result: true };
}

/** Find configFile, function from ts api seems to be broken? */
export function findConfigFile(options: ts.CompilerOptions): ParseResult<string> {
    if (!options.project) {
        return { isValid: false, errorMessage: `error no base path provided, could not find config.`};
    }
    let configPath = options.project;
    // If the project path is wrapped in double quotes, remove them
    if (/^".*"$/.test(configPath)) {
        configPath = configPath.substring(1, configPath.length - 1);
    }
    /* istanbul ignore if: Testing else part is not really possible via automated tests */
    if (!path.isAbsolute(configPath)) {
        // TODO check if options.project can even contain non absolute paths
        configPath = path.join(process.cwd(), configPath);
    }
    if (fs.statSync(configPath).isDirectory()) {
        configPath = path.join(configPath, "tsconfig.json");
    } else if (fs.statSync(configPath).isFile() && path.extname(configPath) === ".ts") {
        // Search for tsconfig upwards in directory hierarchy starting from the file path
        const dir = path.dirname(configPath).split(path.sep);
        for (let i = dir.length; i > 0; i--) {
            const searchPath = dir.slice(0, i).join("/") + path.sep + "tsconfig.json";

            // If tsconfig.json was found, stop searching
            if (ts.sys.fileExists(searchPath)) {
                configPath = searchPath;
                break;
            }
        }
    }

    return { isValid: true, result: configPath };
}
