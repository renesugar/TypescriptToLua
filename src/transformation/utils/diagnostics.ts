import * as ts from "typescript";
import { LuaTarget } from "../../CompilerOptions";
import { createSerialDiagnosticFactory } from "../../utils";
import { AnnotationKind } from "./annotations";

const createDiagnosticFactory = <TArgs extends any[]>(message: string | ((...args: TArgs) => string)) =>
    createSerialDiagnosticFactory((node: ts.Node, ...args: TArgs) => ({
        file: node.getSourceFile(),
        start: node.getStart(),
        length: node.getWidth(),
        messageText: typeof message === "string" ? message : message(...args),
    }));

export const unsupportedNodeKind = createDiagnosticFactory(
    (kind: ts.SyntaxKind) => `Unsupported node kind ${ts.SyntaxKind[kind]}`
);

export const forbiddenForIn = createDiagnosticFactory("Iterating over arrays with 'for ... in' is not allowed.");

export const unsupportedNoSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with a 'this' parameter to function${nameReference} with no 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: void'."
    );
});

export const unsupportedSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with no 'this' parameter to function${nameReference} with 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: any'."
    );
});

export const unsupportedOverloadAssignment = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` to '${name}'` : "";
    return (
        `Unsupported assignment of function with different overloaded types for 'this'${nameReference}. ` +
        "Overloads should all have the same type for 'this'."
    );
});

export const decoratorInvalidContext = createDiagnosticFactory("Decorator function cannot have 'this: void'.");

export const annotationInvalidArgumentCount = createDiagnosticFactory(
    (kind: AnnotationKind, got: number, expected: number) => `'@${kind}' expects ${expected} arguments, but got ${got}.`
);

export const extensionCannotConstruct = createDiagnosticFactory(
    "Cannot construct classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExtend = createDiagnosticFactory(
    "Cannot extend classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExport = createDiagnosticFactory(
    "Cannot export classes with '@extension' or '@metaExtension' annotation."
);

export const extensionInvalidInstanceOf = createDiagnosticFactory(
    "Cannot use instanceof on classes with '@extension' or '@metaExtension' annotation."
);

export const extensionAndMetaExtensionConflict = createDiagnosticFactory(
    "Cannot use both '@extension' and '@metaExtension' annotations on the same class."
);

export const metaExtensionMissingExtends = createDiagnosticFactory(
    "'@metaExtension' annotation requires the extension of the metatable class."
);

export const invalidForRangeCall = createDiagnosticFactory((message: string) => `Invalid @forRange call: ${message}.`);

export const luaTableMustBeAmbient = createDiagnosticFactory(
    "Classes with the '@luaTable' annotation must be ambient."
);

export const luaTableCannotBeExtended = createDiagnosticFactory(
    "Cannot extend classes with the '@luaTable' annotation."
);

export const luaTableInvalidInstanceOf = createDiagnosticFactory(
    "The instanceof operator cannot be used with a '@luaTable' class."
);

export const luaTableCannotBeAccessedDynamically = createDiagnosticFactory("@luaTable cannot be accessed dynamically.");

export const luaTableForbiddenUsage = createDiagnosticFactory(
    (description: string) => `Invalid @luaTable usage: ${description}.`
);

export const luaIteratorForbiddenUsage = createDiagnosticFactory(
    "Unsupported use of lua iterator with '@tupleReturn' annotation in for...of statement. " +
        "You must use a destructuring statement to catch results from a lua iterator with " +
        "the '@tupleReturn' annotation."
);

export const unsupportedAccessorInObjectLiteral = createDiagnosticFactory(
    "Accessors in object literal are not supported."
);

export const unsupportedRightShiftOperator = createDiagnosticFactory(
    "Right shift operator is not supported for target Lua 5.3. Use `>>>` instead."
);

const getLuaTargetName = (version: LuaTarget) => (version === LuaTarget.LuaJIT ? "LuaJIT" : `Lua ${version}`);
export const unsupportedForTarget = createDiagnosticFactory(
    (functionality: string, version: Exclude<LuaTarget, LuaTarget.Universal>) =>
        `${functionality} is/are not supported for target ${getLuaTargetName(version)}.`
);

export const unsupportedProperty = createDiagnosticFactory(
    (parentName: string, property: string) => `${parentName}.${property} is unsupported.`
);

export const invalidAmbientIdentifierName = createDiagnosticFactory(
    (text: string) => `Invalid ambient identifier name '${text}'. Ambient identifiers must be valid lua identifiers.`
);

export const unresolvableRequirePath = createDiagnosticFactory(
    (path: string) => `Cannot create require path. Module '${path}' does not exist within --rootDir.`
);

export const unsupportedVarDeclaration = createDiagnosticFactory(
    "`var` declarations are not supported. Use `let` or `const` instead."
);
