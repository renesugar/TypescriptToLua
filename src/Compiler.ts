import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";

import * as CommandLineParser from "./CommandLineParser";
import {CompilerOptions, LuaLibImportKind, LuaTarget} from "./CompilerOptions";
import {LuaTranspiler} from "./LuaTranspiler";

export function compile(argv: string[]): void {
    const parseResult = CommandLineParser.parseCommandLine(argv);

    if (parseResult.isValid === true) {

        if (parseResult.result.options.help) {
            console.log(CommandLineParser.getHelpString());
            return;
        }

        if (parseResult.result.options.version) {
            console.log(CommandLineParser.version);
            return;
        }

        /* istanbul ignore if: tested in test/compiler/watchmode.spec with subproccess */
        if (parseResult.result.options.watch) {
            watchWithOptions(parseResult.result.fileNames, parseResult.result.options);
        } else {
            compileFilesWithOptions(parseResult.result.fileNames, parseResult.result.options);
        }
    } else {
        console.error(`Invalid CLI input: ${parseResult.errorMessage}`);
    }
}

/* istanbul ignore next: tested in test/compiler/watchmode.spec with subproccess */
export function watchWithOptions(fileNames: string[], options: CompilerOptions): void {
    let host: ts.WatchCompilerHost<ts.SemanticDiagnosticsBuilderProgram>;
    let config = false;
    if (options.project) {
        config = true;
        host = ts.createWatchCompilerHost(options.project, options, ts.sys, ts.createSemanticDiagnosticsBuilderProgram);
    } else {
        host = ts.createWatchCompilerHost(fileNames, options, ts.sys, ts.createSemanticDiagnosticsBuilderProgram);
    }

    host.afterProgramCreate = program => {
        const transpiler = new LuaTranspiler(program.getProgram());

        const status = transpiler.emitFilesAndReportErrors();
        const errorDiagnostic: ts.Diagnostic = {
            category: undefined,
            code: 6194,
            file: undefined,
            length: 0,
            messageText: "Found 0 errors. Watching for file changes.",
            start: 0,
        };
        if (status !== 0) {
            errorDiagnostic.messageText = "Found Errors. Watching for file changes.";
            errorDiagnostic.code = 6193;
        }
        host.onWatchStatusChange(errorDiagnostic, host.getNewLine(), program.getCompilerOptions());
    };

    if (config) {
        ts.createWatchProgram(
            host as ts.WatchCompilerHostOfConfigFile<ts.SemanticDiagnosticsBuilderProgram>
        );
    } else {
        ts.createWatchProgram(
            host as ts.WatchCompilerHostOfFilesAndCompilerOptions<ts.SemanticDiagnosticsBuilderProgram>
        );
    }
}

export function compileFilesWithOptions(fileNames: string[], options: CompilerOptions): void {
    const program = ts.createProgram(fileNames, options);

    const transpiler = new LuaTranspiler(program);

    transpiler.emitFilesAndReportErrors();
}

const libCache: {[key: string]: string} = {};

const defaultCompilerOptions: CompilerOptions = {
    luaLibImport: LuaLibImportKind.Require,
    luaTarget: LuaTarget.Lua53,
};

export function createStringCompilerProgram(
    input: string, options: CompilerOptions = defaultCompilerOptions, filePath = "file.ts"): ts.Program {
    const compilerHost =  {
        directoryExists: () => true,
        fileExists: (fileName): boolean => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: () => "lib.es6.d.ts",
        getDirectories: () => [],
        getNewLine: () => "\n",

        getSourceFile: (filename: string) => {
            if (filename === filePath) {
                return ts.createSourceFile(filename, input, ts.ScriptTarget.Latest, false);
            }
            if (filename.indexOf(".d.ts") !== -1)Â {
                if (!libCache[filename]) {
                    const typeScriptDir = path.dirname(require.resolve("typescript"));
                    const filePath = path.join(typeScriptDir, filename);
                    if (fs.existsSync(filePath)) {
                        libCache[filename] = fs.readFileSync(filePath).toString();
                    } else {
                        const pathWithLibPrefix = path.join(typeScriptDir, "lib." + filename);
                        libCache[filename] = fs.readFileSync(pathWithLibPrefix).toString();
                    }
                }
                return ts.createSourceFile(filename, libCache[filename], ts.ScriptTarget.Latest, false);
            }
            return undefined;
        },

        readFile: () => "",

        useCaseSensitiveFileNames: () => false,
        // Don't write output
        writeFile: (name, text, writeByteOrderMark) => undefined,
    };
    return ts.createProgram([filePath], options, compilerHost);
}

export function transpileString(
    str: string,
    options: CompilerOptions = defaultCompilerOptions,
    ignoreDiagnostics = false,
    filePath = "file.ts"
): string {
    const program = createStringCompilerProgram(str, options, filePath);

    if (!ignoreDiagnostics) {
        const diagnostics = ts.getPreEmitDiagnostics(program);
        const typeScriptErrors = diagnostics.filter(diag => diag.category === ts.DiagnosticCategory.Error);

        if (typeScriptErrors.length > 0) {
            typeScriptErrors.forEach(e => console.warn(e.messageText));
            throw new Error("Encountered invalid TypeScript.");
        }
    }

    const transpiler = new LuaTranspiler(program);

    const result = transpiler.transpileSourceFile(program.getSourceFile(filePath));

    return result.trim();
}
