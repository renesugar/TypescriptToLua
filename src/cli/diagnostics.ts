import * as ts from "typescript";
import { createSerialDiagnosticFactory, createDiagnosticFactoryWithCode } from "../utils";

export const tstlOptionsAreMovingToTheTstlObject = createSerialDiagnosticFactory((tstl: Record<string, any>) => ({
    category: ts.DiagnosticCategory.Warning,
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
}));

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <TArgs extends any[]>(code: number, getMessage: (...args: TArgs) => string) =>
    createDiagnosticFactoryWithCode(code, (...args: TArgs) => ({ messageText: getMessage(...args) }));

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
