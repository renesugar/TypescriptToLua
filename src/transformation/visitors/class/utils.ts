import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { AnnotationKind, getTypeAnnotations } from "../../utils/annotations";

export function isStaticNode(node: ts.Node): boolean {
    return (node.modifiers ?? []).some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendsClause(node: ts.ClassLikeDeclarationBase): ts.HeritageClause | undefined {
    return (node.heritageClauses ?? []).find(clause => clause.token === ts.SyntaxKind.ExtendsKeyword);
}

export function getExtendedNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    const extendsClause = getExtendsClause(node);
    if (!extendsClause) return;

    const superType = context.checker.getTypeAtLocation(extendsClause.types[0]);
    const annotations = getTypeAnnotations(superType);
    if (!annotations.has(AnnotationKind.PureAbstract)) {
        return extendsClause.types[0];
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedNode = getExtendedNode(context, node);
    return extendedNode && context.checker.getTypeAtLocation(extendedNode);
}
