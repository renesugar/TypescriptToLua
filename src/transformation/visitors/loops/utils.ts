import * as ts from "typescript";
import * as lua from "../../../LuaAST";
import { TransformationContext } from "../../context";
import { performHoisting, popScope, pushScope, ScopeType } from "../../utils/scope";
import { isAssignmentPattern } from "../../utils/typescript";
import { transformAssignment } from "../binary-expression/assignments";
import { transformAssignmentPattern } from "../binary-expression/destructuring-assignments";
import { transformBlockOrStatement } from "../block";
import { transformIdentifier } from "../identifier";
import { checkVariableDeclarationList, transformBindingPattern } from "../variable-declaration";

export function transformLoopBody(
    context: TransformationContext,
    loop: ts.WhileStatement | ts.DoStatement | ts.ForStatement | ts.ForOfStatement | ts.ForInOrOfStatement
): lua.Statement[] {
    pushScope(context, ScopeType.Loop);
    const body = performHoisting(context, transformBlockOrStatement(context, loop.statement));
    const scope = popScope(context);
    const scopeId = scope.id;

    if (!scope.loopContinued) {
        return body;
    }

    const baseResult: lua.Statement[] = [lua.createDoStatement(body)];
    const continueLabel = lua.createLabelStatement(`__continue${scopeId}`);
    baseResult.push(continueLabel);

    return baseResult;
}

export function getVariableDeclarationBinding(
    context: TransformationContext,
    node: ts.VariableDeclarationList
): ts.BindingName {
    checkVariableDeclarationList(context, node);

    if (node.declarations.length === 0) {
        return ts.createIdentifier("____");
    }

    return node.declarations[0].name;
}

export function transformForInitializer(
    context: TransformationContext,
    initializer: ts.ForInitializer,
    block: lua.Block
): lua.Identifier {
    const valueVariable = lua.createIdentifier("____value");

    if (ts.isVariableDeclarationList(initializer)) {
        // Declaration of new variable

        const binding = getVariableDeclarationBinding(context, initializer);
        if (ts.isArrayBindingPattern(binding) || ts.isObjectBindingPattern(binding)) {
            block.statements.unshift(...transformBindingPattern(context, binding, valueVariable));
        } else {
            // Single variable declared in for loop
            return transformIdentifier(context, binding);
        }
    } else {
        // Assignment to existing variable(s)

        block.statements.unshift(
            ...(isAssignmentPattern(initializer)
                ? transformAssignmentPattern(context, initializer, valueVariable)
                : transformAssignment(context, initializer, valueVariable))
        );
    }

    return valueVariable;
}
