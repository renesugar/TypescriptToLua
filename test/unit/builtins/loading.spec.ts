import * as tstl from "../../../src";
import { unsupportedProperty } from "../../../src/transformation/utils/diagnostics";
import * as util from "../../util";

describe("luaLibImport", () => {
    test("inline", () => {
        util.testExpression`[0].push(1)`
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Inline })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).not.toContain('require("lualib_bundle")'))
            .expectToMatchJsResult();
    });

    test("require", () => {
        util.testExpression`[0].push(1)`
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Require })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).toContain('require("lualib_bundle")'))
            .expectToMatchJsResult();
    });

    test("always", () => {
        util.testModule``
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Always })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).toContain('require("lualib_bundle")'))
            .expectToEqual(undefined);
    });
});

test.each([tstl.LuaLibImportKind.Inline, tstl.LuaLibImportKind.None, tstl.LuaLibImportKind.Require])(
    "should not include lualib without code (%p)",
    luaLibImport => {
        util.testModule``.setOptions({ luaLibImport }).tap(builder => expect(builder.getMainLuaCodeChunk()).toBe(""));
    }
);

test("lualib should not include tstl header", () => {
    util.testExpression`[0].push(1)`.tap(builder =>
        expect(builder.getMainLuaCodeChunk()).not.toContain("Generated with")
    );
});

describe("Unknown builtin property", () => {
    test("access", () => {
        util.testExpression`Math.unknownProperty`
            .disableSemanticCheck()
            .expectDiagnosticsToMatchSnapshot([unsupportedProperty.code]);
    });
});
