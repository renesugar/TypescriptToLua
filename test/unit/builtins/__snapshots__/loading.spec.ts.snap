// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unknown builtin property access: code 1`] = `
"local ____exports = {}
____exports.__result = Math.unknownProperty
return ____exports"
`;

exports[`Unknown builtin property access: diagnostics 1`] = `"main.ts(1,30): error TSTL: Math.unknownProperty is unsupported."`;
